lib/logdna/client.rb

```diff
@@ -44,95 +50,125 @@ def process_message(msg, opts = {})
+            slices.push(@buffer.pop(pop_point).map { |message| message.source })
+          end
+
+          slices
         else
-          schedule_flush
-        end
-      else
-        @side_message_lock.synchronize do
-          @side_messages.push(process_message(msg, opts))
+          []
         end
       end
-    end
 
-    # This method has to be called with @lock
-    def send_request
-      @side_message_lock.synchronize do
-        @buffer.concat(@side_messages)
-        @side_messages.clear
+      slices.reverse_each do |slice|
+        send_request(slice)
       end
+    end
 
+    def send_request(slice)
+      # TODO: Remove instance-global request object
       @request.body = {
         e: "ls",
-        ls: @buffer
+        ls: slice
       }.to_json
 
-      handle_exception = lambda do |message|
-        @internal_logger.debug(message)
-        @exception_flag = true
-        @side_message_lock.synchronize do
-          @side_messages.concat(@buffer)
-        end
-      end
+      flush_id = Random.uuid
+      error_header = "Flush {#{flush_id}} [#{slice.length} lines] failed."
+      tries = 0
+      loop do
+        tries += 1
 
-      begin
-        @response = Net::HTTP.start(
-          @uri.hostname,
-          @uri.port,
-          use_ssl: @uri.scheme == "https"
-        ) do |http|
-          http.request(@request)
-        end
-        if @response.is_a?(Net::HTTPForbidden)
-          @internal_logger.debug("Please provide a valid ingestion key")
-        elsif !@response.is_a?(Net::HTTPSuccess)
-          handle_exception.call("The response is not successful #{@response}")
+        if tries >= 4
+          @internal_logger.debug("Flush {#{flush_id}} exceeded 3 tries. Discarding flush buffer")
+          break
         end
-        @exception_flag = false
-      rescue SocketError
-        handle_exception.call("Network connectivity issue")
-      rescue Errno::ECONNREFUSED => e
-        handle_exception.call("The server is down. #{e.message}")
-      rescue Timeout::Error => e
-        handle_exception.call("Timeout error occurred. #{e.message}")
-      ensure
-        @buffer.clear
-      end
-    end
 
-    def flush
-      if @lock.try_lock
-        @flush_scheduled = false
-        if @buffer.any? || @side_messages.any?
-          send_request
+        begin
+          response = Net::HTTP.start(
+            @uri.hostname,
+            @uri.port,
+            use_ssl: @uri.scheme == "https"
+          ) do |http|
+            http.request(@request)
+          end
+
+          code = response.code.to_i
+          if code == 401 || code == 403
+            @internal_logger.debug("#{error_header} Please provide a valid ingestion key. Discarding flush buffer")
+            break
+          elsif code == 408 || code == 500 || code == 504
+            # These codes might indicate a temporary ingester issue
+            @internal_logger.debug("#{error_header} The request failed #{response}. Retrying")
```

Comment:
I think there should be a sleep, or a backoff mechanism here, rather than an immediate retry.

Reply:
Agree. I can reimplement this with a backoff and some jitter